Distributed aggregate query

Jadw1 opened the pull request:
This PR extends #9209. It consists of 2 main points:

## UDA reduction and distribution
To enable parallelization of user-defined aggregates, reduction function was added to UDA definition. Reduction function is optional and it has to be scalar function that takes 2 arguments with type of UDA's state and returns UDA's state

## Distribution of native aggregates
All currently implemented native aggregates got their reducible counterpart, which return their state as final result, so it can be reduced with other result. Hence all native aggregates can now be distributed.

# Mixed-cluster testing

## Setup
Local 3-node cluster made with current master. `node1` updated to this branch. Accessing node with `ccm <node-name> cqlsh`

## Test
I've tested belowed things from both old and new node:
- creating UDA with reduce function - not allowed
- selecting count(*) - distributed
- selecting other aggregate function - not distributed

# Issues
Fixes: #10224

raphaelsc commented on db/schema_tables.cc:
```diff
@@ -1469,6 +1497,14 @@ static std::vector<data_type> read_arg_types(const query::result_set_row& row, c
 
+static std::vector<data_value> read_arg_values(const query::result_set_row& row) {
+    std::vector<data_value> arg_values;
+    for (const auto& arg : get_list<sstring>(row, "argument_types")) {
```
for (auto&& arg : ...) then arg can be moved into arg_values.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1512,6 +1540,14 @@ static std::vector<data_type> read_arg_types(const query::result_set_row& row, c
     return arg_types;
 }
 
+static std::vector<data_value> read_arg_values(const query::result_set_row& row) {
```
The function should be better named, it's not clear it is specific to scylla_aggregates.

Jadw1 replied:
I've  based the name on almost identical function just above which has name `read_arg_types` and returns `vector<data_type>`. It also read column named `argument_types`.

avikivity replied:
One badly named function isn't an excuse for another. Also, at least the existing one has a matching function name and column name. The new one is called arg_values but reads a column named argument_types.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1512,6 +1540,14 @@ static std::vector<data_type> read_arg_types(const query::result_set_row& row, c
 
+static std::vector<data_value> read_arg_values(const query::result_set_row& row) {
+    std::vector<data_value> arg_values;
+    for (const auto& arg : get_list<sstring>(row, "argument_types")) {
```
get_list<> already returns a vector.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1567,7 +1603,7 @@ static std::vector<data_type> read_arg_types(const query::result_set_row& row, c
 #endif
 
-static shared_ptr<cql3::functions::user_function> create_func(replica::database& db, const query::result_set_row& row) {
+static shared_ptr<cql3::functions::user_function> create_func(replica::database& db, const query::result_set_row& row, lw_shared_ptr<query::result_set> _rs) {
```
_ prefix is for members, not arguments.

Better name rs to reflect its role, nobody will guess why the function is receiving both a row and a results set.

Jadw1 replied:
I've removed this argument, it's no longer necessary

avikivity commented on db/schema_tables.cc:
```diff
@@ -1638,25 +1674,61 @@ static shared_ptr<cql3::functions::user_aggregate> create_aggregate(replica::dat
+    return ::make_shared<cql3::functions::user_aggregate>(name, initcond, std::move(state_func), std::move(reduce_func), std::move(final_func));
+}
+
+future<lw_shared_ptr<query::result_set>> 
```
I think this should be an optional<row> to simplify use later on.

Jadw1 replied:
This is a good idea but `row` doesn't have no-throw move constructor, so other option is to allocate pointer with `make_unique`, but then allocation is required.

Jadw1 replied:
This function no longer exists, because of using before/after schema

avikivity commented on db/schema_tables.cc:
```diff
@@ -1638,25 +1674,61 @@ static shared_ptr<cql3::functions::user_aggregate> create_aggregate(replica::dat
+    co_await proxy.local().get_db().invoke_on_all([&] (replica::database& db) -> future<> {
         for (const auto& val : diff.created) {
-            cql3::functions::functions::add_function(create(db, *val));
+            auto scylla_specific_ag = co_await extract_scylla_specific_aggregate_info(proxy, *val);
```
Why aren't you getting the row from before/after? It conflicts with how everything else works?

Jadw1 replied:
I was referring to @psarna's [#10144](https://github.com/scylladb/scylla/pull/10144/commits/58529591a9ac9be49f53a69ed33cb9161406be95#diff-2cc928e139ea64f5257e11df02cd2819944fc93210107633315ee2d0edd05c27) with extracting scylla specyfic info for keyspaces.

avikivity replied:
I don't understand.

psarna replied:
I also didn't include `scylla_keyspaces` table into the before/after scheme, because I considered it a little different than other system tables. While for all the tables that describe tables, keyspaces, types, and so on, their system tables actually represent the existence of something in the data dictionary. For these tables, it's very convenient to compare the diff before and after the change, because then we can distinguish if a table or a keyspace has been created, modified or deleted.

For scylla_keyspaces (and also other scylla_* tables) the situation is quite different, because it only stores auxiliary data that depends on existing items only. These tables are technically just another set of columns for existing tables, but we split them to a separate entity to avoid backward compatibility issues. And because of that, it was substantially easier to not take them into account in the already complicated `do_merge_schema` code, which would mean adding parameters to lots of functions and rpc verbs. Instead, the auxiliary information is simply loaded locally later, when we already know what was created, deleted or modified.

I expressed a similar rationale earlier when #10144 was reviewed and the series ultimately got merged, so either someone acked the idea or missed it.

avikivity replied:
I don't understand why it's more complicated to use the same method that we used before, and less complicated to use different methods for different table.

psarna replied:
It makes the changes much less intrusive - for instance, `scylla_tables` was added the old way, which includes fetching and propagating an additional set of mutations, adding it to idl (https://github.com/scylladb/scylla/blob/53c66fe24ae9f25673e0d85467976177bfb2ea03/idl/frozen_schema.idl.hh#L19), and propagating this info via multiple helper functions.

With the new way, one only needs to get the additional data when generating mutations for a particular function, so the code changes are way more local - the basic interface for merging schemas stays the same, and we just generate a few more mutations only for functions that need it. The interface for generating such mutations returns vector<mutation>, so it's already designed to cover mutations from multiple tables. I think it's logically closer to what these scylla_* tables actually represent - it's just auxiliary info that would normally just be added to their corresponding original tables, but we don't do that just to stay compatible with all tools. Then, the new auxiliary info ends up serialized in the same set of mutations as everything else related to functions, and the upper layers, like idl, don't need to know that there's a separate subset of mutations that represent additional UDA information, stored in a separate optional parameter. At the same time, a cluster feature guards this auxiliary info and it will not be propagated unless the whole cluster is aware of it and capable of processing it.

avikivity replied:
What are we losing by not propagating the mutation? I haven't analyzed it deeply but it's very surprising one method that uses RPC to propagate information is the same as another method that queries the data locally.

This is very odd.

avikivity replied:
I still don't like it.

avikivity commented on cql3/selection/abstract_function_selector.cc:
```diff
@@ -79,6 +79,15 @@ abstract_function_selector::new_factory(shared_ptr<functions::function> fun, sha
+                return false;
+            }
+            return p->has_reducefunc();
+        }
```
This should not be specific to UDA but to aggregates in general. SUM() and COUNT() can also have a reducer.

Jadw1 replied:
Changed `has_reducefunc` to `is_reducible` and move this function from UDA to function class

avikivity replied:
Please check that all build-in aggregate functions are now reducible. Otherwise, making them reducible later will require  a cluster feature again.

Jadw1 replied:
Checked. All of them are reducible.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1629,7 +1629,10 @@ std::unique_ptr<prepared_statement> select_statement::prepare(data_dictionary::d
-            && selection->is_count()            // Only count(*) selection is supported.
+            && ( // SUPPORTED PARALLELIZATION
+                selection->is_count() ||
+                selection->is_uda_reducable()
```
So these two conditions are really one.

Jadw1 replied:
Fiexd conditions and separated them by parallelization version (only COUNT(*) / all aggregates)

avikivity commented on conf/scylla.yaml:
```diff
@@ -258,9 +258,10 @@ batch_size_fail_threshold_in_kb: 1024
-# experimental_features:
-#     - udf
+experimental_features:
+    - udf
```
Why this?

Jadw1 replied:
This will be removed when PR will be completed. I needed to rebranch and thus commited this file.

avikivity commented on cql3/functions/user_aggregate.hh:
```diff
@@ -33,7 +33,7 @@ public:
         return *_sfunc;
     }
-    const scalar_function& reducefunc() const {
+    scalar_function& reducefunc() const {
```
Seems wrong. The function itself should not be mutable.

Jadw1 replied:
Fixed

avikivity commented on cql3/selection/selectable.cc:
```diff
@@ -104,6 +105,18 @@ make_count_rows_function_expression() {
+expr::expression
+make_function_expression(const sstring& keyspace, const sstring& name, const std::vector<sstring>& args) {
+    std::vector<expr::expression> args_expr;
+    for(auto& a : args) {
```
for is not a function.

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1348,6 +1349,16 @@ indexed_table_select_statement::find_index_clustering_rows(query_processor& qp,
+    if (selection->is_count()) {
+        return query::forward_request::count{};
+    }
+    else if (selection->is_uda_reducable()) {
```
Our coding style says '} else if {`.

avikivity commented on idl/forward_request.idl.hh:
```diff
@@ -1,9 +1,12 @@
+        sstring keyspace;
+        sstring uda_name;
+        std::vector<sstring> column_names;
     };
```
You can't change IDL in a non-backward-compatible manner, it breaks upgrades.

psarna replied:
We discussed this one briefly, and since distributed COUNT(*) didn't make it to any release so far, I think it's much better to break now, since nobody is going to upgrade from one unofficial commit to another, than to have a weird optional<variant> at the end of this request, which renders the original enum rather redundant. Similarly, this series also piggy-backs on count(*)'s cluster feature for the same reason - to avoid churn since we're not part of any official release.

avikivity replied:
That depends on when 5.1 will be branched and when this series makes it in.

Jadw1 replied:
Now IDL is changed properly, only one new field added at the end, optional and annotated with `[[version 5.1]]`

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -18,6 +18,7 @@
 #include "functions.hh"
 #include "native_aggregate_function.hh"
 #include "exceptions/exceptions.hh"
```
Not reviewing a patch with no explanation.

Jadw1 commented:
I've updated the PR, now all aggregates are parallelized.

Ping @psarna @avikivity

avikivity commented on cql3/Cql.g:
```diff
@@ -664,6 +664,7 @@ createAggregateStatement returns [std::unique_ptr<cql3::statements::create_aggre
         std::vector<shared_ptr<cql3_type::raw>> arg_types;
         std::optional<sstring> ffunc;
         std::optional<expr::expression> ival;
+        std::optional<sstring> rfunc;
```
Changelog should explain why we add reducefunc, not just state that we add it.

Jadw1 replied:
Changed commit's message

avikivity commented on cql3/statements/create_aggregate_statement.cc:
```diff
@@ -42,6 +42,13 @@ shared_ptr<functions::function> create_aggregate_statement::create(query_process
     }
 
+    ::shared_ptr<cql3::functions::scalar_function> reduce_func = nullptr;
+    if(_rfunc) {
```
Coding style says that `if` gets a space afterwards.

avikivity commented on cql3/statements/create_aggregate_statement.cc:
```diff
@@ -42,6 +42,13 @@ shared_ptr<functions::function> create_aggregate_statement::create(query_process
+    if(_rfunc) {
+        reduce_func = dynamic_pointer_cast<functions::scalar_function>(functions::functions::find(functions::function_name{_name.keyspace, _rfunc.value()}, {state_type, state_type}));
+        if (!reduce_func) {
+            throw exceptions::invalid_request_exception(format("Reduce function not found: {}", _rfunc));
```
The problem could be that it wasn't found (for specific parameter types), or that it was not a scalar function. The error can be more helpful.

Jadw1 replied:
Added state type in error message

avikivity commented on db/schema_tables.cc:
```diff
@@ -1566,7 +1599,7 @@ static std::vector<data_type> read_arg_types(replica::database& db, const query:
 #endif
 
-static shared_ptr<cql3::functions::user_function> create_func(replica::database& db, const query::result_set_row& row) {
+static shared_ptr<cql3::functions::user_function> create_func(replica::database& db, const query::result_set_row& row, lw_shared_ptr<query::result_set>) {
```
We now have two result-set style rows. How can we distinguish between the two? They should have names that describe their roles.

Also, one is a ow and another is a result set, which is odd.

avikivity replied:
And that parameter is not used.

Jadw1 replied:
Removed this argument (because of before/after schema)

avikivity commented on db/schema_tables.cc:
```diff
@@ -1637,25 +1670,61 @@ static shared_ptr<cql3::functions::user_aggregate> create_aggregate(replica::dat
+            cql3::functions::functions::find(cql3::functions::function_name{name.keyspace, rfunc}, {state_type, state_type}));
 
+        if (!reduce_func) {
+            throw std::runtime_error(format("Reduce function {} needed by aggregate {} not found", rfunc, name.name));
```
Throwing here is risky. It can end up preventing a node from booting.

Maybe it's better just not to create the aggregate (but log an error instead).

Jadw1 replied:
Because of before/after schema, reduce function is not adding during creation of UDA

avikivity commented on db/schema_tables.cc:
```diff
@@ -2027,11 +2096,29 @@ std::vector<mutation> make_create_aggregate_mutations(shared_ptr<cql3::functions
+    std::vector<mutation> muts = {m};
+
+    if (aggregate->has_reducefunc()) {
+        schema_ptr ss = scylla_aggregates();
```
why is a schema types variable called ss?

Jadw1 replied:
Above schema for aggregates is called `s` and its mutation `p`, so I've added prefix 's' for "scylla". I'll make better name for it

Jadw1 replied:
renamed

avikivity commented on db/schema_tables.cc:
```diff
@@ -2027,11 +2096,29 @@ std::vector<mutation> make_create_aggregate_mutations(shared_ptr<cql3::functions
+
+    if (aggregate->has_reducefunc()) {
+        schema_ptr ss = scylla_aggregates();
+        auto sp = get_mutation(ss, *aggregate);
```
Why is a mutation called sp?

Jadw1 replied:
renamed

avikivity commented on db/schema_tables.hh:
```diff
@@ -112,6 +112,7 @@ static constexpr auto VIEWS = "views";
 static constexpr auto TYPES = "types";
 static constexpr auto FUNCTIONS = "functions";
 static constexpr auto AGGREGATES = "aggregates";
+static constexpr auto SCYLLA_AGGREGATES = "scylla_aggregates";
```
The whole thing needs to be protected with a cluster feature / schema feature so a partially upgraded cluster doesn't break.

avikivity replied:
?

Jadw1 replied:
This is commit that add SCYLLA_AGGREGATES schema feature

https://github.com/scylladb/scylla/pull/10295/commits/cb4b22a869f1db3b61e64d06c5e748a47ac6f04d

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -37,6 +40,16 @@ class impl_count_function : public aggregate_function::aggregate {
     virtual void add_input(cql_serialization_format sf, const std::vector<opt_bytes>& values) override {
         ++_count;
     }
+    virtual void set_accumultor(const opt_bytes& acc) override {
```
spelling

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -37,6 +40,16 @@ class impl_count_function : public aggregate_function::aggregate {
         ++_count;
     }
+    virtual void set_accumultor(const opt_bytes& acc) override {
+        _count = value_cast<int64_t>(long_type->deserialize(bytes_view(*acc)));
```
null check?

Jadw1 replied:
added

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -37,6 +40,16 @@ class impl_count_function : public aggregate_function::aggregate {
+        return long_type->decompose(_count);
+    }
+    virtual void reduce(cql_serialization_format sf, const opt_bytes& acc) override {
+        auto other = value_cast<int64_t>(long_type->deserialize(bytes_view(*acc)));
```
null check?

Jadw1 replied:
added

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -85,12 +101,14 @@ struct accumulator_for : public std::conditional_t<std::is_integral_v<T>,
 
 class impl_user_aggregate : public aggregate_function::aggregate {
     ::shared_ptr<scalar_function> _sfunc;
+    ::shared_ptr<scalar_function> _rfunc;
```
Did you consider using the same approach for non-user aggregates? Defining a separate scalar function, instead of adding new methods for reduction? Maybe it's simpler overall.

Jadw1 replied:
`impl_user_aggregate` also has added new `reduce()` method, which is calling this scalar function, so basically the approach is the same for native aggregates and UDAs. The difference is the UDA has defined reduction function as a scalar function and native aggregate has own implementation of reduce function.

avikivity replied:
We should, in the future, make native aggregates have external state rather than internal accumulators. This will increase uniformity with UDAs.

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -183,15 +225,62 @@ class impl_avg_function_for final : public aggregate_function::aggregate {
+        return tuple_val.serialize();
+    }
+};*/
+
```
What's this?

Will we have both reducible and non-reducible average functions? Why?

Jadw1 replied:
There is need for aggregate query to return its partial result (without applying `compute()` on final accumulator), so the simplest way was to create reducible counterpart of `aggregate_function` which `compute()` returns the accumulator. Other way to achieve this would be to dig into `result_set_builder` and `selectors`.

avikivity replied:
I see.

avikivity replied:
Please add comments explaining this in the aggregate base class.

Jadw1 replied:
Updated commend for `reducible_aggregate_function()` function

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -183,15 +225,62 @@ class impl_avg_function_for final : public aggregate_function::aggregate {
+
+        _sum = value_cast<Type>(tuple[0]);
+        _count = value_cast<int64_t>(tuple[1]);*/
+    }
```
See apache/cassandra@7872318d63009193415ba1365beedc2303a92386. If we make the accumulator an ABI, we have to be careful with how it is implemented, since mistakes are now more costly to fix.

Jadw1 replied:
Could you specify this comment a little more? Would you like to implement AVG state like Cassandra (pair of current avg and counter)?

avikivity replied:
There are two questions:
1. should we switch to Cassandra's implementation?
2. if so, when should we switch? before or after this series?

The answer to the first question if likely yes. Quoting the JIRA:

> For fixed-size integer types overflow is all but guaranteed to happen, yielding incorrect result. While for sum it is somewhat acceptable as the result cannot fit the type, this is not the case for average.

We don't suffer from that because we use a wide accumulator. We should probably keep our current implementation, but if we change the floating-point implementation, we can export and import the accumulator as average+count.

> As the result of average is always within the scope of the source type, failing to produce it only signifies a bad implementation. Yes, one can solve this by type-casting, but do we really want to always have to be telling people that the correct spelling of the average function is cast(avg(cast(value as bigint))) as int), especially if this is so trivial to fix?
>
> Additionally, the straightforward addition we use for floating point versions is not a good choice numerically for larger numbers of values. We should switch to a more stable version, e.g. iterative mean using avg = avg + (value - avg) / count.

We should look for evidence that this is correct.

If we do change, we should change before this series is merged, otherwise making the change later involves a cluster feature and tagging every RPC to indicate which version of the accumulator we use.

Maybe a simple change is to just change the accumulator import/export format to match Cassandra. Then we're insulated from the effects of the change. We also don't have to worry about using varint.

Jadw1 replied:
We've been discussing with @psarna and our conclusion was we probably should stay with sum+count state. 
In the reducible AVG we still need to have sum to reduce two state into one, so in avg+count model this requires multiplying and using wide accumulator to not overflow the sum.

So to really benefit from avg+count implementation the aggregation shouldn't be distributed.

avikivity replied:
> We've been discussing with @psarna and our conclusion was we probably should stay with sum+count state. In the reducible AVG we still need to have sum to reduce two state into one, so in avg+count model this requires multiplying and using wide accumulator to not overflow the sum.

It's no longer an accumulator but a temporary variable. I think it's better not to have to transport varints around.

> 
> So to really benefit from avg+count implementation the aggregation shouldn't be distributed.

What about the the other point about floating-point stability?

Jadw1 replied:
Why transporting varint is a problem? Is it good enough argument to switch to avg+count state only to not transport varint?

Isn't floating-point stability argument also breaking in our case where we have to calculate by multiplying avg * count?

Jadw1 replied:
We can avoid varint  but the intermediate avg has to be big_decimal and transported as `decimal_type`.

So following the sync conversation, should we switch to avg+count?

avikivity replied:
Let's keep the current implementation. I think it's compatible with the more involved algorithm (we just have to multiply the current average with the count before transporting it).

avikivity commented on cql3/functions/aggregate_function.hh:
```diff
@@ -56,6 +63,12 @@ public:
 
+        virtual void set_accumultor(const opt_bytes& acc) = 0;
+
+        virtual opt_bytes get_accumulator() = 0;
```
const?

Jadw1 replied:
const :)

avikivity commented on gms/feature_service.hh:
```diff
@@ -108,6 +108,7 @@ public:
     gms::feature tombstone_gc_options { *this, "TOMBSTONE_GC_OPTIONS"sv };
     gms::feature parallelized_aggregation { *this, "PARALLELIZED_AGGREGATION"sv };
     gms::feature keyspace_storage_options { *this, "KEYSPACE_STORAGE_OPTIONS"sv };
+    gms::feature full_parallelized_aggregation { *this, "FULL_PARALLELIZED_AGGREGATION"sv };
```
It's better to be more specific, because I expect we'll refine it later. What will we call it then, EVEN_MORE_FULL_PARALLELIZED_AGGREGATION?

Also, can fold into the patch that uses it.

Jadw1 replied:
Named it `UDA_NATIVE_PARALLELIZED_AGGREGATION`

avikivity commented on cql3/selection/selector_factories.hh:
```diff
@@ -105,6 +106,22 @@ public:
 
+    bool does_reduction() const {
+        if (_factories.size() != 1) {
+            return false;
```
Why? Does SELECT min(x), max(x) not work?

Jadw1 replied:
I was developing the count(*) parallelization which has this limitation and I wanted to focus on distributing UDAs/native aggregates. Probably it isn't hard to add support to multiple selection, however I wasn't investigating much into this.

avikivity replied:
If we change it later, it will require a schema feature. So better to extend it now, if it's easy.

Jadw1 replied:
Extended

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1626,11 +1626,15 @@ std::unique_ptr<prepared_statement> select_statement::prepare(data_dictionary::d
+            && ( // SUPPORTED PARALLELIZATION
+                selection->is_count()
+                || selection->is_reducible()
+            )
```
Depends on cluster feature too.

avikivity replied:
Shouldn't this happen after IDL/serialization changes?

avikivity replied:
?

avikivity commented on cql3/statements/select_statement.cc:
```diff
@@ -1626,11 +1626,15 @@ std::unique_ptr<prepared_statement> select_statement::prepare(data_dictionary::d
             && group_by_cell_indices->empty()   // No GROUP BY
             // All potential intermediate coordinators must support forwarding
             && db.features().parallelized_aggregation
+            && db.features().full_parallelized_aggregation
```
Ah, it's here, but it's wrong, since now a partially upgraded cluster will lose the previous support for count parallelization. It should apply only to the new feature.

avikivity commented on cql3/selection/selectable.cc:
```diff
@@ -104,6 +104,16 @@ make_count_rows_function_expression() {
 }
 
+expr::expression
+make_function_expression(const ::shared_ptr<functions::function>& function, const std::vector<sstring>& args) {
```
Why sstring parameters and not expression parameters?

avikivity replied:
And it's a function call expression, not a function expression. I don't see why you can't use expr::function_call directly.

avikivity commented on idl/forward_request.idl.hh:
```diff
@@ -1,9 +1,23 @@
     };
-    std::vector<query::forward_request::reduction_type> reduction_types;
+
+    std::vector<std::pair<query::forward_request::reduction_type, std::optional<query::forward_request::aggregation_info>>> reductions;
```
Won't this break compatibility with older versions?

Jadw1 replied:
fixed

avikivity commented on query-request.hh:
```diff
@@ -9,8 +9,12 @@
 
+#include "cql3/functions/function.hh"
+#include "cql3/functions/function_name.hh"
+#include "cql3/functions/aggregate_function.hh"
```
It's not good for query-request to reach into cql/ and pull things from there.

We need to create a new abstraction for functions, and have both cql/ and here depend on it.

Jadw1 replied:
I've moved this to db/, however only minimal required subset of cql's function (`function`, `function_name`, `aggregate_function`). Also there are aliases to old cql types.

avikivity commented on service/forward_service.cc:
```diff
@@ -172,24 +236,36 @@ future<> forward_service::stop() {
+
+        auto& info = reduction.second.value();
+        std::vector<data_type> types;
+        for(auto& name: info.column_names) {
```
`for` is not a function, so it gets a space.

avikivity commented on test/boost/cql_query_test.cc:
```diff
@@ -5063,3 +5063,60 @@ SEASTAR_TEST_CASE(test_parallelized_select_count) {
+            {long_type->decompose(int64_t((value_count - 1) * value_count / 2))}
+        });
+
+        BOOST_CHECK_EQUAL(stat_parallelized + 1, qp.get_cql_stats().select_parallelized);
```
Need some negative checks when we expect a query not to be parallelized.

Jadw1 replied:
added

Jadw1 commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -183,15 +225,62 @@ class impl_avg_function_for final : public aggregate_function::aggregate {
+        // };
+        // return ::make_shared<reducible_avg_function>();
+        
+    }
```
It was not finished when I was updating the PR so I've commented out this part. It will be updated when I push the PR with the changes.

Jadw1 commented:
PR updated with fixed all comments. This is no longer draft PR.
@avikivity I've marked smaller comments as resolved, so it was easier to not omit some by mistake.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/897/

avikivity commented:
What are the changes?

Jadw1 commented:
Besides fixing your comments, I've implemented AVG, fixed added schema feature and made all tests pass.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/926/

avikivity commented:
> Besides fixing your comments, I've implemented AVG, fixed added schema feature and made all tests pass.

I've no idea what my comments are (still less how you fixed them), I'll need an explicit list of changes.

What is "added schema feature" and what does fixing it mean?

Jadw1 commented:
Ok, sorry, didn't know that. Here is the changelog:

- moved part of cql3 into db
https://github.com/scylladb/scylla/pull/10295/commits/472b385cb2f21c02b8a662783e4db250b81fc4df

- implementation of AVG function reduction
(aggregate_fcts.cc)

- null checks in reduce and set_accumulator functions
(aggregate_fcts.cc)

- allowing to distribute only aggregates with simple arguments
(abstract_function_selector.cc, selector.hh, selector_factories.cc. selector_factories.hh)

- removed `make_function_expression`
(selectable-expr.hh, selectable.cc)

- added schema feature `SCYLLA_AGGREGATES` corresponnding to `scylla_aggregates` system table
https://github.com/scylladb/scylla/pull/10295/commits/ea89bfe35a02c4b62798dd84391c747d7dc5f695

- using before/after schema for `scylla_aggregates` table instead of extracting row
(schema_tables.cc)

- changes cluster feature name from `FULL_PARALLELIZED_AGGREGATION` to `UDA_NATIVE_PARALLELIZED_AGGREGATION`
(feature_service.hh)

- fixed way the aggregate functions are obtained. The ideal way would be using `functions::get()`,
because this functions checks for dynamic created aggregates (like `MAX()` for collection type),
however this would require to mock selectors. I'm using `functions::find()` and check  for dynamic aggregates by hand
(forward_service.cc:118)

- added test when query shouldn't be distributed
(cql_query_test.cc)

- fixed compatibility issue in forward_request
(forward_request.idl.hh, query_request.hh)

avikivity commented on cql3/functions/aggregate_function.hh:
```diff
@@ -31,6 +31,13 @@ public:
+     * @return a reducible <code>Aggregate Function</code>.
+     */
+    virtual ::shared_ptr<aggregate_function> reducible_aggregate_function() = 0;
+
```
Please make the comments helpful by explaining the differences. How does an aggregate signal that it is not reducible? By returning null here?

avikivity replied:
Ah, I see is_reducible below. Please indicate this is only legal if is_reducible().

Jadw1 replied:
Added to function's comment

avikivity commented on cql3/functions/aggregate_function.hh:
```diff
@@ -56,6 +63,12 @@ public:
+        virtual opt_bytes get_accumulator() const = 0;
+
+        virtual void reduce(cql_serialization_format sf, const opt_bytes& acc) = 0;
+
```
I think we should move out accumulator storage out of aggregate functions, so we don't need to call new_aggregate. Functions should be pure const functions without state. That's out of scope for this series of course, but would probably have made it simpler if we started with it.

avikivity commented on cql3/functions/function.hh:
```diff
@@ -50,6 +50,13 @@ public:
+     * @return <code>true</code> if the function is reducible, <code>false</code> otherwise.
+     */
+    virtual bool is_reducible() const = 0;
+
```
Better to move to aggregate, no?

Jadw1 replied:
Do you mean to `aggregate_function` or to `aggregate`?

I've followed `is_aggregate()` function which is defined at `function` level, not `aggregate_function`. The difference would be in some places this would require to cast `function` to `aggregate_function` and then check if it is reducible.

avikivity replied:
I can't think of why we'd need to check for reducibility without knowing it's an aggregate. But if you keep it here, rename it to is_reducible_aggregate() to indicate it's not generic for all functions.

Jadw1 replied:
Moved to `aggregate_function`

avikivity commented on cql3/functions/native_aggregate_function.hh:
```diff
@@ -32,6 +32,10 @@ public:
     }
+
+    virtual bool is_reducible() const override {
+        return true;
```
That's too dangerous. Each aggregate that implements reducible_aggregate_function() should override it separately, so we don't accidentally add a non-reducible aggregate and fail here.

Jadw1 replied:
done

avikivity commented on cql3/selection/selector_factories.hh:
```diff
@@ -121,6 +123,36 @@ public:
+
+    // std::vector<std::pair<query::forward_request::reduction_type, >>
+    std::pair<std::vector<query::forward_request::reduction_type>, std::vector<query::forward_request::aggregation_info>>
+     get_reductions() const {
```
bad indent.

avikivity commented on cql3/selection/selector_factories.hh:
```diff
@@ -121,6 +123,36 @@ public:
+    }
+
+    // std::vector<std::pair<query::forward_request::reduction_type, >>
+    std::pair<std::vector<query::forward_request::reduction_type>, std::vector<query::forward_request::aggregation_info>>
```
It's better to use a struct instead of std::pair, so the names guide readers about what's going on.

Jadw1 replied:
changed to struct

avikivity commented on service/forward_service.cc:
```diff
@@ -172,24 +315,45 @@ future<> forward_service::stop() {
         }
+
+        if (!info) {
+            throw std::runtime_error("No aggregation info for reduction type aggregation.");
```
on_internal_error?

avikivity commented on service/forward_service.cc:
```diff
@@ -359,45 +531,52 @@ future<query::forward_result> forward_service::dispatch(query::forward_request r
-                    } else {
-                        result_ = partial_result;
-                    }
+                    co_return co_await seastar::async([this, &req, &result, &tr_state, &dispatcher, vnodes_with_addr = std::move(vnodes_with_addr)] {
```
Also here.

avikivity commented on test/boost/cql_query_test.cc:
```diff
@@ -5064,6 +5064,104 @@ SEASTAR_TEST_CASE(test_parallelized_select_count) {
+        BOOST_CHECK_EQUAL(stat_parallelized, qp.get_cql_stats().select_parallelized);
+    });
+}
+
```
tests for parallelized UDA?

Please add negative tests too for cases that aren't supported.

Jadw1 replied:
tests added

avikivity commented:
Please add information about testing in mixed clusters. Testing should involve creation of reducible UDA, as well as computing old-style (COUNT(*)) and new-style natives on both old and new coordinators.

Jadw1 commented on cql3/functions/aggregate_function.hh:
```diff
@@ -10,71 +10,12 @@
-     * @return a reducible <code>Aggregate Function</code>.
-     */
-    virtual ::shared_ptr<aggregate_function> reducible_aggregate_function() = 0;
-
```
I'll leave it here for now and move it later if there will be more changes to make.

Jadw1 commented:
Rolling update test:
Reduction function can be added to UDA only if whole cluster supports it.
New parallelization method works only if whole cluster supports it.

There is problem with `COUNT(*)` . If the query is sent from older node, the newer one has error on rpc: `got exception while processing a message: std::out_of_range (deserialization buffer underflow)`. I need to dig a bit deeply why this is happening since I've only added one option to enum (it's still fits into `uint8`) and added optional field at the end.

Changelog:
- added description for REDUCEFUNC to cql-extensions.md
- moved commit that add SCYLLA_AGGREGATES schema feature before creation of scylla_aggregates table
- more detailed comment on `reducible_aggregate_function()`
  (aggregate_function.hh)
- moved `is_reducible()` definition from `native_aggregate_function` to specific aggregate implementations
  (native_aggregate_function.hh, aggregate_fcts.cc)
- raplaced `std::pair` with `query::forward_request::reductions_info`
  (selector_factories.hh, selection.cc, selection.hh)
- added distributed and non-distributed UDA tests
  (cql_query_test.cc)
- added cql-pytest for creating UDA with reducefunc 
  (test_uda.py)
- fixed: UDA's reduction function wasn't protected by cluster feature. Now it's under the same feature as 2nd stage of parallelized aggregation since it is intruduced by the same PR

psarna commented:
@Jadw1 you can try inspecting the packets with wireshark in order to see what's different before and after the upgrade, when you send a `SELECT COUNT(*)` request

psarna commented on idl/forward_request.idl.hh:
```diff
@@ -1,15 +1,31 @@
     db::consistency_level cl;
     lowres_clock::time_point timeout;
+
+    std::optional<std::vector<query::forward_request::aggregation_info>> aggregation_infos;
```
you didn't mark that with a version from which it's supported, like it's done here: https://github.com/scylladb/scylla/blob/6c818f86256ba8c098fca3526e0f25232093f0b6/idl/frozen_schema.idl.hh#L21 . That's why you're getting an underflow. Mark the code with `[[5.1]]` and retest - you should be fine.

psarna replied:
And then, when receiving from the network, this parameter will automatically be wrapped in `rpc::optional`, because it's marked as only existing from a specific release.

avikivity replied:
I think the marking is just informational (misinformational in this case) and doesn't affect the generated code.

psarna replied:
@avikivity maybe it used to be only informational, but I see in the IDL compiler code, that it wraps the generated attribute in rpc::optional:
https://github.com/scylladb/scylla/blob/630959bb77873c60b873bdb18afa367a18fc6048/idl-compiler.py#L374-L375

psarna replied:
88f9f2e9d0

Jadw1 replied:
The attribute was enough to fix this.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1083/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1107/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1128/

psarna commented on service/forward_service.cc:
```diff
@@ -370,34 +556,42 @@ future<query::forward_result> forward_service::dispatch(query::forward_request r
-
-                    return *result;
+                [&result, &req, &tr_state] () -> future<query::forward_result> {
+                    forward_aggregates aggrs(req);
```
It sounds a little wasteful to create the forward_aggregates class by copying all the vectors, just in order to use its `with_thread_if_needed` function, and then discard it and recreate the same object again in line 584 to avoid liveness issues... but on the other hand, these vectors are small and this operation is rather cheap compared to everything else that happens on this path, so perhaps it's good enough to leave it as is.

psarna replied:
Oh wait, no, it can be easily done. At the beginning of this function (line 536) you use `do_with` anyway. That's where you can also create the forward_aggregates, right? Since it only needs a request to be initialized, and you have `req` available - e.g. you could create `forward_aggregates aggrs(req)` right before line 536, and then pass the forward aggregates instance to do_with, so that lives long enough. And then you can pass them via reference to `with_thread_if_needed`.

Jadw1 replied:
I've wrapped `parallel_for_each`'s continuation with `do_with`. I cannot pass forward aggregates to `do_with` you mention, since separate forward_aggregates are needed in `parallel_for_each` (line 544).

avikivity replied:
Why not a coroutine?

Jadw1 replied:
There was a coroutine, but I've made a mistake like one described here https://github.com/scylladb/scylla/pull/10602#issuecomment-1166631581, so I prefer to stay with continuations to avoid this

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1134/

avikivity commented:
I see a lot of comments marked "resolved", but I have no idea what the resolutions were.

Jadw1 commented on db/schema_tables.cc:
```diff
@@ -1638,25 +1674,61 @@ static shared_ptr<cql3::functions::user_aggregate> create_aggregate(replica::dat
     }
+    ::shared_ptr<cql3::functions::scalar_function> reduce_func = nullptr;
+    if (rs && !rs->empty()) {
+        auto&& rrow = rs->row(0);
```
No longer necessary

Jadw1 commented on cql3/selection/selector_factories.hh:
```diff
@@ -105,6 +106,22 @@ public:
+            return std::nullopt;
+        }
+
+        return _factories[0]->get_reduction();
```
returning a vector

Jadw1 commented on cql3/Cql.g:
```diff
@@ -697,13 +698,16 @@ createAggregateStatement returns [std::unique_ptr<cql3::statements::create_aggre
       K_SFUNC sfunc = allowedFunctionName
       K_STYPE stype = comparatorType
+      (
+        K_REDUCEFUNC reduce_name = allowedFunctionName { rfunc = reduce_name; }
```
added

Jadw1 commented on db/schema_features.hh:
```diff
@@ -23,6 +23,7 @@ enum class schema_feature {
     CDC_OPTIONS,
     PER_TABLE_PARTITIONERS,
     SCYLLA_KEYSPACES,
+    SCYLLA_AGGREGATES,
```
moved this commit

Jadw1 commented on service/forward_service.cc:
```diff
@@ -201,20 +365,28 @@ future<query::forward_result> forward_service::dispatch_to_shards(
-                mapped.merge(*partial, req.reduction_types);
-            }
-            return {mapped};
+    return seastar::async([this, req = std::move(req), tr_info = std::move(tr_info)] () mutable {
```
done

Jadw1 commented:
@avikivity I've added comments to each conversation and unresolved them. Only left trivial ones (like coding style or indentation) as resolved

avikivity commented on cql3/functions/aggregate_fcts.cc:
```diff
@@ -497,6 +498,7 @@ bool user_aggregate::is_pure() const { return _sfunc->is_pure() && (!_finalfunc
 bool user_aggregate::is_native() const { return false; }
 bool user_aggregate::is_aggregate() const { return true; }
 bool user_aggregate::requires_thread() const { return _sfunc->requires_thread() || (_finalfunc && _finalfunc->requires_thread()); }
+void user_aggregate::set_reducefunc(::shared_ptr<scalar_function> reducefunc) { _reducefunc = reducefunc; }
```
Why do we need this in addition to the constructor? It's not bad per se, but in general function objects should be immutable so we don't have to worry about updates and invalidation.

avikivity commented on db/schema_tables.cc:
```diff
@@ -1668,6 +1706,48 @@ static future<> merge_aggregates(distributed<service::storage_proxy>& proxy, sch
+            update_reducefunc(db, *val);
+        }
+        for (const auto& val : diff.altered) {
+            update_reducefunc(db, *val);
```
Isn't this creating a window where the reduce function is bad? After merge_functions() is called but before mesge_scylla_aggregates() is called.

If so we should fold merge_scylla_aggregates into merge_functions so it happens simultaneously. This will allow us to drop the mutating set_reducefunc.

avikivity replied:
Merge merge_scylla_aggregates into merge_aggregates().

Nor sure what you meant be "really simultaneously".

Jadw1 replied:
I've joined the diff from both tables, however I've ignored the altered rows for now. (more detailed comment in changelog)

avikivity commented on db/schema_tables.cc:
```diff
@@ -1668,6 +1706,48 @@ static future<> merge_aggregates(distributed<service::storage_proxy>& proxy, sch
+            update_reducefunc(db, *val);
+        }
+        // Skip diff.dropped because corresponding UDA has already been dropped
+        // and drop from scylla_aggregates doesn't drop scalar functions itself
```
And prevent this oddness.

avikivity commented on serializer_impl.hh:
```diff
@@ -771,6 +771,19 @@ std::variant<T...> deserialize(Input& in, boost::type<std::variant<T...>> v) {
+template<typename Output, typename T, typename U>
+void serialize(Output& out, const std::pair<T, U>& p) {
+    serialize(out, p.first);
+    serialize(out, p.second);
```
I think it's better to use structs rather than pairs. First, pairs are hostile to anyone reading the code. Second (pun unintended), IDL structs can be extended while pairs cannot.

Jadw1 replied:
I'll drop this commit. I was using pair when I've changed IDL in the bad way, breaking the compatibility. Now it's not used.

avikivity commented on query-request.hh:
```diff
@@ -373,7 +376,7 @@ struct forward_request {
     };
     struct aggregation_info {
-        cql3::functions::function_name name;
+        db::functions::function_name name;
```
Shouldn't this happen in the patch that changed the namespace?

Jadw1 replied:
Moved

avikivity commented on service/forward_service.cc:
```diff
@@ -42,12 +47,165 @@
+                    auto& arg = types[0];
+                    if (arg->is_collection() || arg->is_tuple() || arg->is_user_type()) {
+                        func = cql3::functions::aggregate_fcts::make_count_rows_function();
+                    }
```
Can't we make the functions code do all this? It seems wrong to have the service know about all the functions.

Jadw1 replied:
Moved to `cql3::functions::functions::find_with_dynamic_aggr()`

avikivity commented on service/forward_service.cc:
```diff
@@ -200,21 +379,28 @@ future<query::forward_result> forward_service::dispatch_to_shards(
-    auto result = co_await container().map_reduce0(
-        [req, tr_info] (auto& fs) {
+    for (const auto& s : smp::all_cpus()) {
+        futures.push_back(container().invoke_on(s, [req, tr_info] (auto& fs) {
```
Why the change from map_reduce?

Jadw1 replied:
because of the `with_thread_if_needed` required and reduce function in `map_reduce0` cannot return future

avikivity commented:
Please add information about mixed-clustering testing to the cover-letter.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1250/

avikivity commented on db/schema_tables.cc:
```diff
@@ -1075,6 +1106,10 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, std:
+    auto old_scylla_aggregates = schema_result();
+    if (proxy.local().features().cluster_schema_features().contains<schema_feature::SCYLLA_KEYSPACES>()) {
+        old_scylla_aggregates = co_await read_schema_for_keyspaces(proxy, SCYLLA_AGGREGATES, keyspaces);
+    }
```
This can (should?) be unconditional. schema_features are about what we can export to other nodes. Locally we're always self-consistent - if the code has support for the new table, then it exists.

I can't think of a concrete reason to make it unconditional (if the cluster hasn't agreed on the schema feature, then the table should be empty), it just seems less fragile to me, to avoid races during boot.

btw, why are you checking schema_featrure::SCYLLA_KEYSPACES for the SCYLLA_AGGREGATES table?

Jadw1 replied:
fixed

avikivity commented on db/schema_tables.cc:
```diff
@@ -1096,14 +1131,18 @@ static future<> do_merge_schema(distributed<service::storage_proxy>& proxy, std:
+    auto new_scylla_aggregates = schema_result();
+    if (proxy.local().features().cluster_schema_features().contains<schema_feature::SCYLLA_KEYSPACES>()) {
+        new_scylla_aggregates = co_await read_schema_for_keyspaces(proxy, SCYLLA_AGGREGATES, keyspaces);
+    }
```
same here.

avikivity replied:
but you should check the feature when exporting schema mutations to send to another node.

avikivity commented on db/schema_tables.cc:
```diff
@@ -2198,7 +2198,7 @@ std::vector<mutation> make_create_aggregate_mutations(schema_features features,
 std::vector<mutation> make_drop_aggregate_mutations(schema_features features, shared_ptr<cql3::functions::user_aggregate> aggregate, api::timestamp_type timestamp) {
     auto muts = make_drop_function_mutations(aggregates(), *aggregate, timestamp);
-    if (features.contains<schema_feature::SCYLLA_AGGREGATES>() && aggregate->has_reducefunc()) {
+    if (features.contains<schema_feature::SCYLLA_AGGREGATES>() && aggregate->is_reducible()) {
```
Would have been less noisy if the original patch called it is_reducible().

Jadw1 replied:
fixed

avikivity commented on cql3/functions/functions.cc:
```diff
@@ -360,6 +360,56 @@ functions::find(const function_name& name, const std::vector<data_type>& arg_typ
+// mock or serialize expressions and `functions::find()` is not enough,
+// because it does not search for dynamic aggregate functions
+shared_ptr<function>
+functions::find_with_dynamic_aggr(const function_name &name, const std::vector<data_type> &arg_types) {
```
Our coding style has `&` sticking to the type, not the argument.

I don't understand the function. What does "with_dynamic_aggr" mean? I could try to reverse-engineer the meaning from the code, but the function name should explain itself better.

avikivity replied:
If it's copied from function::get(), maybe an extra parameter to function::get() is indicated, or sharing of the code with a helper.

Jadw1 replied:
renamed and extracted the common part

avikivity commented:
Looks better, very close now.

Jadw1 commented:
Changelog:
- renamed `find_with_dynamic_aggr` to `mock_get` (cause it is basically mocking get functionality and should be removed when we will have selector serialization) and extracted common part of get to helper function

- removed `has_reducefunc` to `is_reducible` in commit that adds reducer to UDA

- https://github.com/scylladb/scylla/pull/10295#discussion_r917390236
fixed this comment

- fixed bugs that causing failing tests

Jadw1 commented on idl/forward_request.idl.hh:
```diff
@@ -1,15 +1,31 @@
     };
-    std::vector<query::forward_request::reduction_type> reduction_types;
+
+    std::vector<query::forward_request::reduction_type> reductions;
```
reverted

avikivity commented:
@annastuchlik I don't see a page that describes used-defined aggregate functions under docs/. Maybe it's because they're still under the experimental flag.

annastuchlik commented:
> @annastuchlik I don't see a page that describes used-defined aggregate functions under docs/. Maybe it's because they're still under the experimental flag.

@avikivity The feature was introduced in 4.6, but it wasn't documented. I only found the information about it in the release notes for 4.6. I think it should be documented - and marked as experimental. We can remove the label when it's no longer experimental. Soon the docs will be versioned, so we can make updates per release - and easily document experimental features.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1363/

avikivity commented:
> Mixed-cluster testing
> 
> I've tested belowed things from both old and new node:
> 
> * creating UDA with reduce function - not allowed
> * selecting count(*) - distributed
> * selecting other aggregate function - not distributed

Please add this to the cover letter (and add information about how you ensured both old and new coordinators were accessed)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1410/

avikivity commented:
> For now, due to disability to serialize select statement, only single selection is supported.


This is outdated, yes?

Jadw1 commented:
> > For now, due to disability to serialize select statement, only single selection is supported.
> 
> This is outdated, yes?

Yes

Jadw1 commented:
> > Mixed-cluster testing
> > I've tested belowed things from both old and new node:
> > 
> > * creating UDA with reduce function - not allowed
> > * selecting count(*) - distributed
> > * selecting other aggregate function - not distributed
> 
> Please add this to the cover letter (and add information about how you ensured both old and new coordinators were accessed)

Cover letter updated
